sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -4296,15 +4296,23 @@ pi_result piProgramBuild(pi_program Program, pi_uint32 NumDevices,
   ze_device_handle_t ZeDevice = DeviceList[0]->ZeDevice;
   ze_context_handle_t ZeContext = Program->Context->ZeContext;
   ze_module_handle_t ZeModule = nullptr;
-  ZE_CALL(zeModuleCreate, (ZeContext, ZeDevice, &ZeModuleDesc, &ZeModule,
-                           &Program->ZeBuildLog));
-
+  ze_result_t ZeResult =
+      ZE_CALL_NOCHECK(zeModuleCreate, (ZeContext, ZeDevice, &ZeModuleDesc,
+                                       &ZeModule, &Program->ZeBuildLog));
+  if (ZeResult != ZE_RESULT_SUCCESS) {
+    // We need to clear Program state to avoid double destroy of zeModule in
+    // case where SYCL RT calls piProgramRelease().
+    // We should not return with an error code here due to the comments below.
+    Program->ZeModule = nullptr;
+    Program->Code.reset();
+    Program->State = _pi_program::Invalid;
+  }
   // The call to zeModuleCreate does not report an error if there are
   // unresolved symbols because it thinks these could be resolved later via a
   // call to zeModuleDynamicLink.  However, modules created with piProgramBuild
   // are supposed to be fully linked and ready to use.  Therefore, do an extra
   // check now for unresolved symbols.
-  ze_result_t ZeResult = checkUnresolvedSymbols(ZeModule, &Program->ZeBuildLog);
+  ZeResult = checkUnresolvedSymbols(ZeModule, &Program->ZeBuildLog);
   if (ZeResult != ZE_RESULT_SUCCESS) {
     // remove ZeModule that is associated with the failed program
     ZE_CALL(zeModuleDestroy, (ZeModule));
```

Comment:
This should be removed

Reply:
instead set Program->State = _pi_program::Invalid;

Reply:
We need to call zeModuleDestroy() to fix the memory leak.
We can't assume the user will catch the exception thrown by SYCL RT.
We destroy the ZeModule here and we won't doubly destroy it because we set Program->ZeModule as nullptr, right?

Reply:
Users must catch exceptions or their program is terminated and no cleanup is done (by design)
If users do catch exception then the program destructor will be called and the L0 module will be destroyed from piProgramRelease.

Reply:
> Users must catch exceptions or their program is terminated and no cleanup is done (by design)
We can't twist our user's arm to catch the exception always.
It is not our control.
> If users do catch exception then the program destructor will be called and the L0 module will be destroyed from piProgramRelease.
No, we already destroyed it when we detect the build failure.
Additionally, we set Program->ZeModule=nullptr, so even if RT calls piProgramRelease(), L0 won't call zeModuleDestroy() for this Program.

Reply:
> We can't twist our user's arm to catch the exception always.

And we don't need to. If users are comfortable with the fact their process is terminated, then we do not guarantee (C++ does not guarantee actually) proper unwinding, and destruction. The memory "leaked" will be reclaimed by OS.

> we already destroyed it when we detect the build failure.

My point is that we should not destroy it if build failed. It is responsibility of piProgramRelease to destroy it.

Reply:
But, RT does not know the internal details of L0 plugin.
L0 plugin create ZeModule as a part of program build, and I believe it is more natural to cleanup whatever L0 plugin created internally.
Please note that this issue is about memory leak in `ZeModule` not `pi_program`.

Reply:
> RT does not know the internal details of L0 plugin.

Right. All it needs to do is call PI program create/build/release
The PI (and L0) cleanup is performed in piProgramRelease